File: crates/yaca-plugin/src/protocol.rs
//! JSON-RPC 2.0 wire frames over newline-delimited stdio (mirrors
//! `yaca_mcp::protocol`).

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const JSONRPC_VERSION: &str = "2.0";

pub mod codes {
    pub const METHOD_NOT_FOUND: i64 = -32601;
    pub const INVALID_PARAMS: i64 = -32602;
    pub const INTERNAL_ERROR: i64 = -32603;
    /// App-defined: a guard hook vetoed the action.
    pub const VETO: i64 = 1;
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct JsonRpcRequest {
    pub jsonrpc: String,
    pub id: u64,
    pub method: String,
    #[serde(default)]
    pub params: Value,
}

impl JsonRpcRequest {
    #[must_use]
    pub fn new(id: u64, method: impl Into<String>, params: Value) -> Self {
        Self {
            jsonrpc: JSONRPC_VERSION.to_string(),
            id,
            method: method.into(),
            params,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct JsonRpcResponse {
    pub jsonrpc: String,
    pub id: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<JsonRpcError>,
}

impl JsonRpcResponse {
    #[must_use]
    pub fn ok(id: u64, result: Value) -> Self {
        Self {
            jsonrpc: JSONRPC_VERSION.to_string(),
            id,
            result: Some(result),
            error: None,
        }
    }

    #[must_use]
    pub fn err(id: u64, code: i64, message: impl Into<String>) -> Self {
        Self {
            jsonrpc: JSONRPC_VERSION.to_string(),
            id,
            result: None,
            error: Some(JsonRpcError {
                code,
                message: message.into(),
                data: None,
            }),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct JsonRpcError {
    pub code: i64,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub data: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct JsonRpcNotification {
    pub jsonrpc: String,
    pub method: String,
    #[serde(default)]
    pub params: Value,
}

impl JsonRpcNotification {
    #[must_use]
    pub fn new(method: impl Into<String>, params: Value) -> Self {
        Self {
            jsonrpc: JSONRPC_VERSION.to_string(),
            method: method.into(),
            params,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Frame {
    Request(JsonRpcRequest),
    Response(JsonRpcResponse),
    Notification(JsonRpcNotification),
}

impl Frame {
    /// # Errors
    /// Returns the parse / classification error message on malformed input.
    pub fn parse(line: &str) -> Result<Self, String> {
        let value: Value = serde_json::from_str(line).map_err(|e| e.to_string())?;
        let obj = value.as_object().ok_or("frame is not a JSON object")?;
        let is_request = obj.contains_key("method") && obj.contains_key("id");
        let is_notification = obj.contains_key("method") && !obj.contains_key("id");
        let is_response = obj.contains_key("result") || obj.contains_key("error");
        if is_request {
            serde_json::from_value(value)
                .map(Frame::Request)
                .map_err(|e| e.to_string())
        } else if is_notification {
            serde_json::from_value(value)
                .map(Frame::Notification)
                .map_err(|e| e.to_string())
        } else if is_response {
            serde_json::from_value(value)
                .map(Frame::Response)
                .map_err(|e| e.to_string())
        } else {
            Err("frame is neither request, response, nor notification".to_string())
        }
    }
}

File: crates/yaca-plugin-opencode/adapter/src/protocol.ts
import { z } from "zod"

export const JSONRPC_VERSION = "2.0"

export const ERROR_CODES = {
  METHOD_NOT_FOUND: -32601,
  INVALID_PARAMS: -32602,
  INTERNAL_ERROR: -32603,
} as const

const JsonRpcRequestSchema = z
  .object({
    jsonrpc: z.literal(JSONRPC_VERSION),
    id: z.number().int().nonnegative(),
    method: z.string().min(1),
    params: z.unknown().optional().default({}),
  })
  .strict()

export type JsonRpcRequest = z.infer<typeof JsonRpcRequestSchema>

export type ParseRequestResult =
  | { readonly ok: true; readonly request: JsonRpcRequest }
  | { readonly ok: false; readonly message: string }

export function parseJsonRpcRequest(line: string): ParseRequestResult {
  let value: unknown
  try {
    value = JSON.parse(line)
  } catch (error) {
    if (error instanceof SyntaxError) {
      return { ok: false, message: error.message }
    }
    if (error instanceof Error) {
      return { ok: false, message: error.message }
    }
    throw error
  }
  const parsed = JsonRpcRequestSchema.safeParse(value)
  if (!parsed.success) {
    return { ok: false, message: parsed.error.message }
  }
  return { ok: true, request: parsed.data }
}

export function okResponse(id: number, result: unknown): string {
  return `${JSON.stringify({ jsonrpc: JSONRPC_VERSION, id, result })}\n`
}

export function errorResponse(id: number, code: number, message: string): string {
  return `${JSON.stringify({
    jsonrpc: JSONRPC_VERSION,
    id,
    error: { code, message },
  })}\n`
}
